internals/overlord/pairingstate/manager.go

```diff
@@ -0,0 +1,312 @@
+	// handshake transcript signature using the public key in this
+	// certificate.
+	roots := x509.NewCertPool()
+	roots.AddCert(clientCert)
+	opts := x509.VerifyOptions{
+		Roots: roots,
+		// We only support verifying client TLS certificates.
+		KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+	}
+	_, err := clientCert.Verify(opts)
+	if err != nil {
+		return fmt.Errorf("cannot verify client certificate signature: %w", err)
+	}
+
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	existingIdentities := m.state.Identities()
+
+	for _, identity := range existingIdentities {
+		if identity.Cert == nil || identity.Cert.X509 == nil {
+			// Not valid certificate identity.
+			continue
+		}
+
+		if identity.Cert.X509.Equal(clientCert) {
+			// This identity is already added so in this special
+			// case we complete the pairing request without adding
+			// it again with a new username.
+			m.paired = true
+			m.state.Set(pairedStateKey, m.paired)
+
+			return nil
+		}
+	}
+
+	username, err := generateUniqueUsername(existingIdentities)
+	if err != nil {
+		return fmt.Errorf("cannot create new identity username: %w", err)
+	}
+
+	newIdentity := &state.Identity{
+		Access: state.AdminAccess,
+		Cert:   &state.CertIdentity{X509: clientCert},
+	}
+
+	err = m.state.AddIdentities(map[string]*state.Identity{
+		username: newIdentity,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot add identity: %w", err)
+	}
+
+	m.paired = true
+	m.state.Set(pairedStateKey, m.paired)
+
+	return nil
+}
+
+// PairingWindowEnabled returns whether the pairing window is currently enabled.
+func (m *PairingManager) PairingWindowEnabled() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.enabled
+}
+
+// stopTimer stops the current timer if it exists
+func (m *PairingManager) stopTimer() {
+	if m.timer != nil {
+		m.timer.Stop()
+	}
+}
+
+// startTimer starts a new timer that will disable the pairing window after
+// the given timeout. If the timer exists, we will reuse the timer by
+// resetting it with a new duration.
+func (m *PairingManager) startTimer(timeout time.Duration) {
+	// If timer already exists, just reset it with the new timeout
+	if m.timer != nil {
+		m.timer.Reset(timeout)
+		return
+	}
+
+	m.timer = timeAfterFunc(timeout, func() {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		m.enabled = false
+	})
+}
+
+// EnablePairing requests the pairing manager to enable the pairing window.
+func (m *PairingManager) EnablePairing(timeout time.Duration) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// If the pairing window is already enabled, reset the timeout duration
+	if m.enabled {
+		m.startTimer(timeout)
+		return nil
+	}
+
+	// Check the pairing mode
+	switch m.config.Mode {
+	case ModeDisabled, ModeUnset:
+		return errors.New("cannot enable pairing with pairing mode disabled")
+
+	case ModeSingle:
+		// Single mode: check if already paired
+		if m.paired {
+			return errors.New("cannot enable pairing when already paired in 'single' pairing mode")
+
+		}
```

Comment:
Bogus empty line. The comment right above is also unnecessary, as it's just reading the surrounding code.

Reply:
Fixed, thanks. Resolving.